### ISSUE TITLE
Ink ascent/descent

### ISSUE THREAD

**Issue by fred-wang** [w3c/mathml-core#78]
This one is a bit tricky... It was mentioned in w3c/mathml#43 and https://github.com/mathml-refresh/mathml/issues/52#issuecomment-514016715 at least. But it needs clear explanation.

When drawing a box, one can distinguish its normal size (the space it occupies) from its "ink" size (smallest enclosing rect for the painted content). For example, a white space has nonzero width/height but an empty ink box.

First the ink size is involved in the OpenType MATH spec when computing vertical gaps [1]. If I'm correct, it is not involved for horizontal placement during layout of math box, so I think we can don't need "ink left/right bearings" and can focus on ink ascent/descent below.

Using CSS baseline concepts, the MathML Core specification defines two new baselines "ink-over baseline" and "ink-under baseline" which delimitate the ink ascent/descent of a box [2]. All the boxes/baselines in the SVG schemas of the spec are drawn with dashes but the ink boxes are suggested with a background. 

The status in browser is the following:
* All MathML boxes in Gecko have ink metrics, both vertical and horizontal.
* WebKit does not use ink metrics. For token elements, it calculate box metrics of the text specially [3] which means that it will somewhat work with basic constructions.
* For Blink, @rwlbuis started to implement the concept of the core spec but it's still very experimental and incomplete. It at least somewhat works for text and fractions.

In order to understand this properly, we need a non-trial test case. I uploaded a test case [4] with basic text/fraction/square root examples and the interesting one which is AAA/√CCC (square root as a denominator).

The font is built like this [5]: 
* all font ascents/descents are 1em and the "use typo metrics" bit is set.
* Glyphs for A, B, C are 1em squares (respectively 1em ink ascent, 1em ink descent and .5em ink ascent/descent). They are respectively drawn in red, green and blue.
* Glyph for U+221a (radical symbol) is 1em width, 1em ink ascent, 1em ink descent. It can be repeated with 1em overlap to build a glyph assembly.
* fraction and radical rulethickness are 1em
* num/denom min gaps are 1em
* num/denom shifts are 2em
* radical vertical gap is 1em
* radical extra ascender is 2em

Exercise: Deduce the vertical positions of AAA, fraction bar, radical overbar and CCC.

Gecko, WebKit, Blink, XeLaTeX and LuaLaTeX render this last example inconsistently. Analysis is complicated by the fact that there are some unrelated bugs and that I was not able to make XeLaTeX/LuaLaTeX use my custom glyphs for the text.

Two questions for the CG:

(1) Is it important to have this concept of ink ascent/descent for all boxes? Or only male ascent/descent of token elements match the text ink ascent/descent? As I understand the MATH spec, it is important to have it for all boxes...

(2) If it is important, is the MathML Core description correct? For example, is the "radical vertical gap" included in its ink box or not? I guess we should ask Microsoft people about this. 

[1] https://docs.microsoft.com/en-us/typography/opentype/spec/math
[2] https://mathml-refresh.github.io/mathml-core/#box-model
[3] https://trac.webkit.org/browser/webkit/trunk/Source/WebCore/css/mathml.css#L19
[4] https://github.com/mathml-refresh/ink-metrics
[5] https://github.com/mathml-refresh/ink-metrics/blob/master/create-font.py

**Comment by fred-wang** [comment]
To elaborate a bit more concretely, this is how browsers render the following example: 

    <p><math><mtext style="background: pink">−</mtext></math></p>
    <p><math><mtext style="background: lightgreen">_</mtext></math></p>
    <p><math><mtext style="background: lightblue">_−</mtext></math></p>

Gecko calculate both "normal" and ink metrics for all MathML boxes, so it's possible to properly positioned math stuff with ink metrics, while still using "normal" metrics for the background.

WebKit and Blink with WIP patch make normal and ink metrics match for token elements, so background may not be visible (or be very close the ink extent of the text).

Blink without the patch uses only normal metrics, so the background remains large but there are too many gaps in MathML formulas.

![mtext-with-minus-and-underscore](https://user-images.githubusercontent.com/567455/156340021-ba500580-5a0a-4b79-8fcf-119e2c0d01b3.png)

**Comment by davidcarlisle** [comment]
@fred-wang  thanks for the nice explanation and image...

Looking just at your image in the last comment the Firefox one looks best but if implementation wise it adds a lot of complication it may be complication we might live without.  Certainly in TeX usage people often don't like boxes that too closely mirror individual character height/depth as you end up with very ragged highlighting, and so you often end up with things like `\colorbox{\strut .}`  \colorbox{\strut A}`   where the `\strut` forces any content within a reasonable size range to have a minimal height and depth, so the boxed . and boxed A have the same vertical extents.   If end users are explicitly hiding the character height to get even sized background boxes, then the finer points of how the default box size is obtained from the metrics may (??) not be so important.

That said the webkit one that is (from somewhere?) always showing a small part of the background seems to be harder to explain but may cause less confused end users than the blink+patch version, I can see many people asking where their backgrounds have gone. But it is I think explainable that the background is matching (and obscured by) the character,

**Comment by davidcarlisle** [comment]
@fred-wang  the second image is definitely much better and if that causes a few features about background colouring that need to be carefully explained I'd still say it looks like a win. The blink rendering is really starting to look nice. I'd support your initial implementation plan here.

**Comment by clapierre** [comment]
> ```
> <p><math><mtext style="background: pink">−</mtext></math></p>
> <p><math><mtext style="background: lightgreen">_</mtext></math></p>
> <p><math><mtext style="background: lightblue">_−</mtext></math></p>
> ```

Be careful, it is not recommended to have inline styling as it can not be overridden by a person who may be color blind or need to adjust these colors, font sizes etc.

could the same thing be done instead with CSS?
IE:

```
<p><math><mtext class="background_pink">−</mtext></math></p>
<p><math><mtext class="background_lightgreen">_</mtext></math></p>
<p><math><mtext class="background_lightblue">_−</mtext></math></p>

[CSS]
.background_pink {background-color: pink;}
.background_lightgreen {background-color: lightgreen; }
.background_lightblue {background-color: lightblue;}
```

**Comment by fred-wang** [comment]
@clapierre yes sure, putting CSS in a separate style sheet is valid in MathML. I only used inline style attribute to make the example less verbose.

**Comment by dginev** [comment]
re Fred's comment https://github.com/w3c/mathml-core/issues/78#issuecomment-1056874643
> Notice that the former contains large vertical gaps, which is definitely not desired.

I've stumbled on the flip side of that issue, where there is no gap at all, in the root case. Fred cross-referenced this issue in my Chromium report, so back-referencing that report here: https://bugs.chromium.org/p/chromium/issues/detail?id=1395834

I'll inline the image for convenience:
<img alt="radical signs holding content without top padding" src="https://user-images.githubusercontent.com/348975/211199388-caa8ca57-a60e-490f-a024-8242693c0fa5.png" width="600px">

I'm nowhere near informed enough to make any useful technical comments on how MathML Core should be improved here.

**Comment by Jamesernator** [comment]
I'm not sure if all of these problems are due to this cause, but it looks like the same issue. Basically pretty much all fonts (*including the default*) have weird behaviours with `<msqrt>` in the Chromium implementation:

Top and bottom of sqrts are not aligned:

![sqrt-underlines-and-overlines](https://user-images.githubusercontent.com/8518303/221406528-801a8e4d-44d8-4c1d-b9af-506a48315d69.png)

Large gaps under sqrt are common:

![sqrt-gaps-2](https://user-images.githubusercontent.com/8518303/221406570-c0b9cd6f-bce9-4e75-a03c-472ebfbeb15e.png)

The radical fails to line up with the overbar extender:

![sqrt-overbar-misalign](https://user-images.githubusercontent.com/8518303/221406586-360bc408-97d0-4f81-9c0f-31c84ef4ec92.png)

Both MathJax and Firefox (I have not checked Safari) have far more reasonable behaviour with their respective default fonts:

MathJax:

![sqrt-mathjax](https://user-images.githubusercontent.com/8518303/221406685-ec7508ee-3de6-41f6-b898-ca1d16bb65c8.png)

Firefox:

![sqrt-firefox](https://user-images.githubusercontent.com/8518303/221406700-97c9b52e-d071-4698-8a8d-5fb704809c88.png)
